File: read_data_using_schema.scala
import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.{functions => F}

object Reading_file {

val fireSchema = StructType(
                    Array(
                         StructField("CallNumber", IntegerType, true),
                         StructField("UnitID", StringType, true),
                         StructField("IncidentNumber", IntegerType, true),
                         StructField("CallType", StringType, true),
                         StructField("CallDate", StringType, true),
                         StructField("WatchDate", StringType, true),
                         StructField("CallFinalDisposition", StringType, true),
                         StructField("AvailableDtTm", StringType, true),
                         StructField("Address", StringType, true),
                         StructField("City", StringType, true),
                         StructField("Zipcode", IntegerType, true),
                         StructField("Battalion", StringType, true),
                         StructField("StationArea", StringType, true),
                         StructField("Box", StringType, true),
                         StructField("OriginalPriority", StringType, true),
                         StructField("Priority", StringType, true),
                         StructField("FinalPriority", IntegerType, true),
                         StructField("ALSUnit", BooleanType, true),
                         StructField("CallTypeGroup", StringType, true),
                         StructField("NumAlarms", IntegerType, true),
                         StructField("UnitType", StringType, true),
                         StructField("UnitSequenceInCallDispatch", IntegerType, true),
                         StructField("FirePreventionDistrict", StringType, true),
                         StructField("SupervisorDistrict", StringType, true),
                         StructField("Neighborhood", StringType, true),
                         StructField("Location", StringType, true),
                         StructField("RowID", StringType, true),
                         StructField("Delay", FloatType, true)
                    )
                        )
val fireCalls = "../data/fire-calls.csv"
val fireDF = spark.read.schema(fireSchema).option("header" , true).csv(fireCalls)
fireDF.show()

// ------------------------------------------------------------------------
//                       Saving and in many formats..
// Save it as parquet
val parquetPath = "any path"
fireDF.write.format("parquet").save(parquetPath)

// Save it as Parquet Table
val parquetTable = "any path"
fireDF.write.format("parquet").saveAsTable(parquetTable)
// ------------------------------------------------------------------------
//                  Performing transformations and actions
val fewFireDF = fireDF.select("CallNumber" , "UnitID" , "IncidentNumber").where($"CallType" =!= "Medical Incident")
fewFireDF.show(5 , false)

//show distinct call types
fireDF.select("CallType").where(col("CallType").isNotNull).agg(countDistinct('CallType) as 'DistinctCallTypes).show()

//filter for only distinct non-null CallTypes from all the rows
fireDF.select("CallType").where(col("CallType").isNotNull).distinct().show(10, false)

//rename  Delay column..
val newFireDF = fireDF.withColumnRenamed("Delay" , "ResposneDelayedinMins")

// select delays longer than 5 
newFireDF.select("ResposneDelayedinMins").where($"ResposneDelayedinMins" > 10).show()

val fireTsDF = newFireDF.withColumn("IncidentDate", to_timestamp(col("CallDate"), "MM/dd/yyyy")).drop("CallDate")
.withColumn("OnWatchDate", to_timestamp(col("WatchDate"), "MM/dd/yyyy")).drop("WatchDate")
.withColumn("AvailableDtTS", to_timestamp(col("AvailableDtTm"),"MM/dd/yyyy hh:mm:ss a")).drop("AvailableDtTm")
// Select the converted columns
fireTsDF.select("IncidentDate", "OnWatchDate", "AvailableDtTS").show()
fireTsDF.select(year($"IncidentDate")).distinct().orderBy(year($"IncidentDate")).show()

// the most common types of fire calls
fireTsDF
.select("CallType")
.where(col("CallType").isNotNull)
.groupBy("CallType")
.count()
.orderBy(desc("count"))
.show(10, false)

}

File: read_data_using_schema.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
import pyspark.sql.functions as F

fire_schema = StructType([
    StructField("CallNumber", IntegerType(), True),
    StructField("UnitID", StringType(), True),
    StructField('IncidentNumber', IntegerType(), True),
    StructField('CallType', StringType(), True),
    StructField('CallDate', StringType(), True),
    StructField('WatchDate', StringType(), True),
    StructField('CallFinalDisposition', StringType(), True),
    StructField('AvailableDtTm', StringType(), True),
    StructField('Address', StringType(), True),
    StructField('City', StringType(), True),
    StructField('Zipcode', IntegerType(), True),
    StructField('Battalion', StringType(), True),
    StructField('StationArea', StringType(), True),
    StructField('Box', StringType(), True),
    StructField('OriginalPriority', StringType(), True),
    StructField('Priority', StringType(), True),
    StructField('FinalPriority', IntegerType(), True),
    StructField('ALSUnit', BooleanType(), True),
    StructField('CallTypeGroup', StringType(), True),
    StructField('NumAlarms', IntegerType(), True),
    StructField('UnitType', StringType(), True),
    StructField('UnitSequenceInCallDispatch', IntegerType(), True),
    StructField('FirePreventionDistrict', StringType(), True),
    StructField('SupervisorDistrict', StringType(), True),
    StructField('Neighborhood', StringType(), True),
    StructField('Location', StringType(), True),
    StructField('RowID', StringType(), True),
    StructField('Delay', FloatType(), True)
])


# Use the dataFrameReader ..
sf_fire_file = "../Projects/Reading_from_files_using_schema/data/fire-calls.csv"
fire_df = spark.read.csv(sf_fire_file, header=True, schema=fire_schema)
fire_df.show()

# ----------------------------------------------------------------------------
#                      Saving and writing in many formats
# Save it as parquet
parquet_path = "any path"
fire_df.write.format("parquet").save(parquet_path)

# Save it as Parquet Table
parquet_table = "any path"
fire_df.write.format("parquet").saveAsTable(parquet_table

# ----------------------------------------------------------------------------
#               Performing transformations and actions
few_frie_df=(fire_df.select("CallNumber", "UnitID", "IncidentNumber").where(
    col("CallType") != "Medical Incident"))

few_frie_df.show(5, truncate=False)

(fire_df
.select("CallType").where(col("CallType").isNotNull())
.alias("DistinctCallTypes").agg(countDistinct("CallType")).show())

# filter for only distinct non-null CallTypes from all the rows
(fire_df
.select("CallType")
.where(col("CallType").isNotNull()).distinct().show(10, False))


new_fire_df=fire_df.withColumnRenamed("Delay", "ResponseDelayedinMins")
(new_fire_df.select("ResponseDelayedinMins").where(
    col("ResponseDelayedinMins") > 5).show(5, False))

# Convert to more usable formats
fire_ts_df=(new_fire_df
.withColumn("IncidentDate", to_timestamp(col("CallDate"), "MM/dd/yyyy"))
.drop("CallDate")
.withColumn("OnWatchDate", to_timestamp(col("WatchDate"), "MM/dd/yyyy"))
.drop("WatchDate")
.withColumn("AvailableDtTS", to_timestamp(col("AvailableDtTm"),
"MM/dd/yyyy hh:mm:ss a"))
.drop("AvailableDtTm"))
# Select the converted columns
fire_ts_df.select("IncidentDate", "OnWatchDate",
                  "AvailableDtTS").show(5, False)
fire_ts_df.select(year('IncidentDate')).distinct().orderBy(
    year("IncidentDate")).show(10, False)

# the most common types of fire calls
(fire_ts_df.select("CallType")
.where(col("CallType").isNotNull()).groupBy("CallType")
.count().orderBy("count", ascending=False)
.show(n=10, truncate=False))

# some Computatinos
(fire_ts_df
.select(F.sum("NumAlarms"), F.avg("ResponseDelayedinMins"),
F.min("ResponseDelayedinMins"), F.max("ResponseDelayedinMins"))
.show())
# ----------------------------------------------------------------------------
